src/compiler/utilities.ts

```diff
@@ -322,6 +322,18 @@ namespace ts {
     }
 
     /**
+     * Note: it is expected that the `nodeArray` and the `node` are within the same file.
+     * For example, searching for a `SourceFile` in a `SourceFile[]` wouldn't work.
+     */
+    export function indexOfNode(nodeArray: ReadonlyArray<Node>, node: Node) {
+        return binarySearch(nodeArray, node, compareNodePos);

```

Comment:
One place this wouldn't work is with synthesized nodes. Not sure if those cases can pop up where it's been used so far.